File: typescript/tests/giftcard/application/integration/payment/handlers/confirmTopUpFromPayment.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import {
  paymentAcceptedIntegrationEvent,
  paymentExpiredIntegrationEvent,
  paymentRejectedIntegrationEvent,
} from '@/payment/integration/paymentResultIntegrationEvent.js';
import { ConfirmTopUpFromPayment } from '@/giftcard/application/integration/payment/handlers/confirmTopUpFromPayment.js';
import { TopUpConfirming } from '@/giftcard/application/usecases/topUpConfirming.js';
import { PaymentResult } from '@/giftcard/application/integration/payment/adapter/paymentResult.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../../../../testsupport/databaseSetup.js';
import { GiftCardAggregateFactory } from '../../../../../testsupport/giftcard/aggregateFactory.js';

describe('ConfirmTopUpFromPayment', () => {
  let repository: SqliteGiftCardRepository;
  let confirmTopUpFromPayment: ConfirmTopUpFromPayment;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('giftcard');
    repository = new SqliteGiftCardRepository(database);
    const paymentResult = new PaymentResult();
    const useCase = new TopUpConfirming(repository);
    confirmTopUpFromPayment = new ConfirmTopUpFromPayment(paymentResult, useCase);
  });

  describe('construction', () => {
    it('rejects null parameters', () => {
      const paymentResult = new PaymentResult();
      const useCase = new TopUpConfirming(repository);

      expect(() => new ConfirmTopUpFromPayment(null as unknown as PaymentResult, useCase)).toThrow();
      expect(() => new ConfirmTopUpFromPayment(paymentResult, null as unknown as TopUpConfirming)).toThrow();
    });
  });

  describe('payment results handling', () => {
    it('on accepted should update balance', () => {
      const issued = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const amount = new Money(37.25);

      confirmTopUpFromPayment.handle(
        paymentAcceptedIntegrationEvent(issued.id().value.value, amount)
      );

      const updated = repository.findById(issued.id());
      expect(updated).not.toBeNull();
      expect(updated!.balance()).toEqual(amount);
    });

    it('on rejected should do nothing', () => {
      const issued = GiftCardAggregateFactory.getSavedGiftCard(repository);

      confirmTopUpFromPayment.handle(
        paymentRejectedIntegrationEvent(issued.id().value.value, new Money(50), 'declined')
      );

      const persisted = repository.findById(issued.id());
      expect(persisted).not.toBeNull();
      expect(persisted!.balance()).toEqual(Money.zero());
    });

    it('on expired should do nothing', () => {
      const issued = GiftCardAggregateFactory.getSavedGiftCard(repository);

      confirmTopUpFromPayment.handle(
        paymentExpiredIntegrationEvent(issued.id().value.value, new Money(50))
      );

      const persisted = repository.findById(issued.id());
      expect(persisted).not.toBeNull();
      expect(persisted!.balance()).toEqual(Money.zero());
    });

    it('should fail if card does not exist', () => {
      const nonExisting = generateId((value) => new GiftCardId(value));
      const amount = new Money(10);

      expect(() =>
        confirmTopUpFromPayment.handle(
          paymentAcceptedIntegrationEvent(nonExisting.value.value, amount)
        )
      ).toThrow();
    });
  });
});

File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/integration/payment/handlers/ConfirmTopUpFromPaymentTest.java
package it.giannibombelli.wsc2026.giftcard.application.integration.payment.handlers;

import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.integration.payment.adapter.PaymentResult;
import it.giannibombelli.wsc2026.giftcard.application.usecases.TopUpConfirming;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.giftcard.infrastructure.SqliteGiftCardRepository;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;
import it.giannibombelli.wsc2026.testsupport.AggregateFactory;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ConfirmTopUpFromPaymentTest {

    private GiftCardRepository repository;
    private ConfirmTopUpFromPayment confirmTopUpFromPayment;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("giftcard");
        repository = new SqliteGiftCardRepository(dataSource);

        PaymentResult paymentResult = new PaymentResult();
        TopUpConfirming useCase = new TopUpConfirming(repository);
        confirmTopUpFromPayment = new ConfirmTopUpFromPayment(paymentResult, useCase);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullParameters() {
            PaymentResult paymentResult = new PaymentResult();
            TopUpConfirming useCase = new TopUpConfirming(repository);

            assertThatThrownBy(() -> new ConfirmTopUpFromPayment(null, useCase))
                .isInstanceOf(NullPointerException.class);

            assertThatThrownBy(() -> new ConfirmTopUpFromPayment(paymentResult, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class PaymentResultsHandling {
        @Test
        void onAccepted_shouldUpdateBalance() {
            GiftCard issued = AggregateFactory.getSavedGiftCard(repository);
            Money amount = new Money(new BigDecimal("37.25"));

            confirmTopUpFromPayment.handle(
                new PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent(issued.id().value().toString(), amount)
            );

            Optional<GiftCard> updated = repository.findById(issued.id());
            assertThat(updated).isPresent();
            assertThat(updated.get().balance()).isEqualTo(amount);
        }

        @Test
        void onRejected_shouldDoNothing() {
            GiftCard issued = AggregateFactory.getSavedGiftCard(repository);

            confirmTopUpFromPayment.handle(
                new PaymentResultIntegrationEvent.PaymentRejectedIntegrationEvent(
                    UUID.randomUUID().toString(), new Money(new BigDecimal("50.00")), "declined")
            );

            GiftCard persisted = repository.findById(issued.id()).orElseThrow();
            assertThat(persisted.balance()).isEqualTo(Money.zero());
        }

        @Test
        void onExpired_shouldDoNothing() {
            GiftCard issued = AggregateFactory.getSavedGiftCard(repository);

            confirmTopUpFromPayment.handle(
                new PaymentResultIntegrationEvent.PaymentExpiredIntegrationEvent(
                    UUID.randomUUID().toString(), new Money(new BigDecimal("50.00")))
            );

            GiftCard persisted = repository.findById(issued.id()).orElseThrow();
            assertThat(persisted.balance()).isEqualTo(Money.zero());
        }

        @Test
        void shouldFailIfCardDoesNotExist() {
            GiftCardId nonExisting = EntityId.generate(GiftCardId::new);
            Money amount = new Money(new BigDecimal("10.00"));

            assertThatThrownBy(() -> confirmTopUpFromPayment.handle(
                new PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent(nonExisting.value().toString(), amount)
            ))
                .isInstanceOf(IllegalStateException.class);
        }
    }
}
